server/rtcd.go

```diff
@@ -0,0 +1,113 @@
+	var cfg rtcd.ClientConfig
+	p.metrics.IncStoreOp("KVGet")
+	data, appErr := p.API.KVGet(rtcdConfigKey)
+	if appErr != nil {
+		return cfg, fmt.Errorf("failed to get rtcd config: %w", appErr)
+	}
+	if len(data) == 0 {
+		return cfg, nil
+	}
+	if err := json.Unmarshal(data, &cfg); err != nil {
+		return cfg, fmt.Errorf("failed to unmarshal rtcd config: %w", err)
+	}
+	return cfg, nil
+}
+
+func (p *Plugin) getRTCDClientConfig(rtcdURL string) (rtcd.ClientConfig, error) {
+	var cfg rtcd.ClientConfig
+
+	// Give precedence to environment to override everything else.
+	cfg.ClientID = os.Getenv("CALLS_RTCD_CLIENT_ID")
+	cfg.AuthKey = os.Getenv("CALLS_RTCD_AUTH_KEY")
+	cfg.URL = rtcdURL
+	if rtcdURL = os.Getenv("CALLS_RTCD_URL"); rtcdURL != "" {
+		cfg.URL = rtcdURL
+	}
+
+	// Parsing the URL in case it's already containing credentials.
+	if cfg.ClientID == "" && cfg.AuthKey == "" {
+		u, clientID, authKey, err := parseURL(cfg.URL)
+		if err != nil {
+			return cfg, fmt.Errorf("failed to parse URL: %w", err)
+		}
+		cfg.ClientID = clientID
+		cfg.AuthKey = authKey
+		cfg.URL = u
+	}
+
+	// if no URL has been provided until now we fail with error.
+	if cfg.URL == "" {
+		return cfg, fmt.Errorf("rtcd URL is missing")
+	}
+
+	// If no client id has been provided until now we fail with error.
+	if cfg.ClientID == "" {
+		return cfg, fmt.Errorf("client id is missing")
+	}
+
+	// If the auth key is set we proceed and return the config.
+	// Otherwise we need to either fetch the config from the k/v store
+	// or register the client.
+	if cfg.AuthKey != "" {
+		return cfg, nil
+	}
+
+	client, err := rtcd.NewClient(cfg)
+	if err != nil {
+		return cfg, fmt.Errorf("failed to create rtcd client: %w", err)
+	}
+	defer client.Close()
+
+	// Here we need some coordination to avoid multiple plugin instances to
+	// register at the same time (at most one would succeed).
+	mutex, err := cluster.NewMutex(p.API, "rtcd_registration")
+	if err != nil {
+		return cfg, fmt.Errorf("failed to create cluster mutex: %w", err)
+	}
+
+	lockCtx, cancelCtx := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelCtx()
+	if err := mutex.LockWithContext(lockCtx); err != nil {
+		return cfg, fmt.Errorf("failed to acquire cluster lock: %w", err)
+	}
+	defer mutex.Unlock()
+
+	if storedCfg, err := p.getStoredRTCDConfig(); err != nil {
+		return cfg, fmt.Errorf("failed to get rtcd credentials: %w", err)
+	} else if storedCfg.URL == cfg.URL && storedCfg.ClientID == cfg.ClientID {
+		return storedCfg, nil
+	}
+
+	cfg.AuthKey, err = client.Register(cfg.ClientID)
+	if err != nil {
+		return cfg, fmt.Errorf("failed to register rtcd client: %w", err)
+	}
+
+	cfgData, err := json.Marshal(&cfg)
+	if err != nil {
+		return cfg, fmt.Errorf("failed to marshal rtcd client config: %w", err)
+	}
+
+	if err := p.API.KVSet(rtcdConfigKey, cfgData); err != nil {
+		return cfg, fmt.Errorf("failed to store rtcd client config: %w", err)
+	}
```

Comment:
To note, a limitation of the current design, if the application were to crash after the registration succeeded but before we were able to store the config (containing the authkey) the client would be permanently locked out until either the clientID changed or the client was deregistered, just like a user forgetting its own password.

Reply:
Maybe send an `ack` that signals the registration is complete and the authkey can be activated?

Reply:
If we go to that complexity (having to keep additional state and coordination) I am thinking we may as well let the client provide the auth key, At the end of the day it's just like a password. We could still validate it properly on server side to make sure it has enough entropy.

Reply:
Since there's no real authentication happening here, maybe we should just do that? (have the client state their auth key, have the server record it) I'm not really sure what this is doing, security-wise, to be honest. You kind of brought it up in the call, but I wasn't totally convinced by the answer.

Reply:
I guess the advantage was to keep things simple having a central place that generates these keys. Now we are trusting clients to provide them which means we need to add some validation in place. But from a security perspective I don't think it's any different if implemented properly. I can bring this up today so we can converge on a decision asap.

Reply:
@oh6hay Wondering if you could weight on this from a security perspective. Is there any advantage in letting the server (rtcd) generate the authentication key (current implementation) vs having the client (plugin) proposing one, assuming some checks are in place to make sure the key is "strong" enough?